### ISSUE TITLE
`eleventyConfig.augmentFunctionContext` to add `this.eleventy` and `this.page` to the context of any function (but mostly for filters/shortcodes)

### ISSUE THREAD

**Issue by zachleat** [11ty/eleventy#3355]
Sample usage here creating a Nunjucks shortcode from scratch: https://github.com/11ty/eleventy/blob/26d26b7a4fe1e8a7a9431283ff437998a5468cf4/test/EleventyTest.js#L1435-L1445

Related to #3310 and https://github.com/noelforte/eleventy-plugin-vento/issues/9

**Comment by zachleat** [comment]
Fixed by 26d26b7a4fe1e8a7a9431283ff437998a5468cf4

**Comment by pauleveritt** [comment]
This might be a really useful thing for my TSX stuff.

**Comment by zachleat** [comment]
@pauleveritt Yeah absolutely!

**Comment by zachleat** [comment]
Shipped with 3.0.0-alpha.15 https://github.com/11ty/eleventy/releases/tag/v3.0.0-alpha.15

**Comment by noelforte** [comment]
Revisiting this, [with the addition of an `exports` field in eleventy's package.json](https://github.com/11ty/eleventy/blob/7eb257d2a7f795a7eeb518ba2361158f803bed87/package.json#L11-L14), it's no longer possible to access any utilites by their direct filepath. For instance, this throws an error:

```js
import { augmentKeys } from '@11ty/eleventy/src/Engines/Util/ContextAugmenter.js';
```

Does it make sense to mark `ContextAugmenter.js`'s exports as public so they can be consumed elsewhere outside of `eleventyConfig`?

---

As an aside, I just came across #3422 which makes the suggestion to do away with usage of `this` in JavaScript templates. It got me curious if there's a future for Eleventy to maintain some sort of global object containing useful data (like `page` and `eleventy`) that could be imported project-wide, which would alleviate my struggles with `this` and my template syntax plugin for Vento.

**Comment by mayank99** [comment]
I think this might solve some of the problems I was facing which led me to open #3422, but I don't see any docs for `augmentFunctionContext`. Maybe this issue needs the https://github.com/11ty/eleventy/labels/needs-documentation label?

**Comment by zachleat** [comment]
Temporary docs preview deploying here https://11ty-website-git-v3-11ty.vercel.app/docs/languages/custom/#compile

**Comment by noelforte** [comment]
Thanks for this @zachleat! After getting more educated on the implementation for my use case in `eleventy-plugin-vento`. I realized it would be simpler to precompile templates with a direct call to the shortcode instead of an indirect call.

I did stumble upon one thing though, I see `augmentObject` and `augmentKeys` are exported from `src/Engines/Util/ContextAugmenter.js`, but trying to import them throws an error. Filed in #3460, which might also help clean up things.
